src/librustc_mir/interpret/memory.rs

```diff
@@ -791,6 +791,27 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
         self.get_raw(ptr.alloc_id)?.read_c_str(self, ptr)
     }
 
+    /// Reads a 0x0000-terminated u16-sequence from memory. Returns them as a Vec<u16>.
+    /// Terminator 0x0000 is not included in the returned Vec<u16>.
+    ///
+    /// Performs appropriate bounds checks.
+    pub fn read_wide_str(&self, ptr: Scalar<M::PointerTag>) -> InterpResult<'tcx, Vec<u16>> {
+        let mut ptr = self.force_ptr(ptr)?; // We need to read at least 1 byte, so we *need* a ptr.
+        let allocation = self.get_raw(ptr.alloc_id)?;
+        let mut u16_seq = Vec::new();
+
+        loop {
+            let single_u16 = allocation.read_scalar(self, ptr, Size::from_bytes(2))?.to_u16()?;
```

Comment:
This seems almost right now!

The remaining issue is: if you read the `read_scalar` docs, you will see that they say

> It is the caller's responsibility to check bounds and alignment beforehand.

So, you should call `check_ptr_access` before calling `read_scalar`.

In other words, your function docs say "Performs appropriate bounds checks" but they fail do deliver on that promise.